src/Core/Aetherling/Passes/ThroughputModifications.hs

```diff
@@ -1,339 +1,508 @@
+-- parallelize
+-- ASSUMPTION: the user has specified a type for these ops and passes will not
+-- change the type because that would change the semantics of the
+-- program. For example, one could speed up an Add T_Int my making it a
+-- Add $ T_Array 2 T_Int, but that would change the program's meaning.
+attemptSpeedUp requestedMult op@(Add _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(Sub _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(Mul _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(Div _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(Max _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(Min _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(Ashr _ _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(Shl _ _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(Abs _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(Not _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(And _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(Or  _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(XOr _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@Eq = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@Neq = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@Lt = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@Leq = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@Gt = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@Geq = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(LUT _) = (MapOp requestedMult op, requestedMult)
+
+-- Must map over these because making a single memory wider is a more
+-- expensive operation than banking it (making multiple copies)
+-- NOTE: unclear what expensive means here, just my understanding from
+-- talking with Spatial team, Ross, and other hardware people. I should
+-- figure it out.
+-- maping over a MemRead/MemWrite is banking it, and wrapping
+-- the type with an array is making a single memory read/write more
+-- per clock.
+attemptSpeedUp requestedMult op@(MemRead _) = (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(MemWrite _) = (MapOp requestedMult op, requestedMult)
+
+-- These are leaf, non-modifiable rate unlike SequenceArrayRepack because,
+-- for these operations, the user has not specified the type separately from the
+-- array length. Therefore, speedup may modify the meaning of the program
+-- by changing the types of these ops.
+attemptSpeedUp requestedMult op@(ArrayReshape _ _) =
+  (MapOp requestedMult op, requestedMult)
+attemptSpeedUp requestedMult op@(DuplicateOutputs _ _) =
```

Comment:
This is probably not the right place to put this suggestion, but we could consider whether we should change the name of Duplicate to Replicate, since it's not necessarily 2x.